pretyping/evarsolve.ml

```diff
@@ -1628,7 +1628,7 @@ let rec invert_definition unify flags choose imitate_defs
         (* Note that we don't need to declare ?evk'' yet: it may remain virtual *)
         let aliases = lift_aliases k aliases in
         (try
-          let ev = (evk,List.map (lift k) argsv) in
+          let ev = normalize_evar !evdref (evk,List.map (lift k) argsv) in
```

Comment:
This patch is weird, then next function does take the evarmap and passes it down to the API it calls.
IMO we should look for the real bug downstream, one of these functions must be bugged.
Normalizing here is just hiding the bug, not solving it.

Reply:
The important value here is the evar itself, not the term built out of the evar. I am not surprised we need to do this.

Reply:
I'm not sure that there is a "real" bug. The point is that the original `ev` has been restricted while imitating. So, it does not seem wrong to replace it with its instance before trying to project. What kind of "real" bug would you suspect otherwise?

Reply:
Yes some *other* piece of code is received the evar map and not dereferencing it.

It is arbitrary to apply the substitution here. I think @herbelin should try to push it down and see where things really change, otherwise it is the same "technique" to introduce the new proof engine, normalize at goal enter time... not ideal.

Reply:
(and yes, porting the code to EConstr would solve it by typing)

Reply:
AFAIK the problem is find_undefined at https://github.com/coq/coq/blob/477ac28ad5721ce84d5ac5d9afecbfa8c3863314/pretyping/evarsolve.ml#L1073
how would this be done with econstr?

Reply:
@gares we're precisely in the low-level parts of unification, I don't think it makes sense to say we can port this to EConstr. At some point you do have to handle pointers and stuff when implementing your runtime. It must not be exposed to the end user, but it's still a reality in the implementation.

Reply:
Maybe put in another way, `normalize_evar sigma ev` is the same as `destEvar (EConstr.kind sigma (mkEvar ev))` where `destEvar` and `mkEvar` have no choices than being in the low-level part.

I think you also get a point that the fact that `ev` has been restricted on the fly also means that what has been computed from `ev` (namely the `subst`) should be recomputed taking into account the restriction. (But this is a bit more work, so probably better for later.)

Reply:
I see that, but I think I disagree about the comments to EConstr.

The "bug" is that instead of passing around a term (which is quotiented over sigma) we pass an evar key and its arguments:
1) Maybe I was wrong, and the client of the whole procedure should pass a NON instantiated evar.
2) Or maybe the client shall pass a FLEXIBLE term.

Both would be documentable APIs. In both cases it seems too arbitrary to apply the substitution there.
It seems like being resilient to a miscall (in the first case) or using the wrong datatype (in the second case).

Reply:
We are here solving equations of the form `?ev@{args} == term`. On the paper, this would be a stepwise instantiation, i.e.:
1. if `term` can be extracted by projecting some contents from one of the `args`, this gives a possible solution
2. another possibility is to copy the head constructor of `term` at the head of `?ev` and to recursively solve equations on the subterms

In practice, to avoid generating too much evars, the heuristics is to restrict 1. to only variables and applied constructors in the first case and to do 2. in a "big step" way, that is to directly instantiate `?ev(rels,constrs) := term(rels,constrs)` in one go.

I don't see clearly why, but when building `term(rels,constrs)`, `?ev` can be restricted on the fly.

So, I agree that because of this possibility of restriction, the procedure could consider `?ev@{args}` not as a fixed argument of the recursion over `term`, but as a parameter of the loop which takes dynamically into account a possible restriction, if some arrives, so that we maintain the invariant that `?ev@{args}` is a non-instantiated evar.

Regarding the PR, I unfortunately don't have much time to work on a different fix, but that someone experiments a variant of the algorithm which maintains the invariant that `ev` is always a non-instantiated evar would indeed be nice. Preferably, this should go in another PR, so that #15366 is not delayed to much.

Reply:
I'm fine with a hot fix